torch/distributed/_state_dict_utils.py

```diff
@@ -403,11 +410,19 @@ def tensor_func(
         _: Any,
     ) -> torch.Tensor:
         if len(obj.size()) == 0:
-            return torch.tensor(0, dtype=obj.dtype)
+            return torch.tensor(0, dtype=obj.dtype) 
 
         if share_memory:
-            t = torch.empty(*tuple(obj.size()), dtype=obj.dtype).share_memory_()
+            t = torch.empty(*tuple(obj.size()), dtype=obj.dtype)
+            t = t.share_memory_()
             if pin_memory:
+                def unpin_memory(t):
+                    succ = torch.cuda.cudart().cudaHostUnregister(t.data_ptr())
+                    assert (
```

Comment:
Any thoughts on leaving this as a warning? If this is on program exit we may not care

Reply:
I am noob here - but if we ignore the error here a subsequent tensor allocation with caching allocation is likely to error out. This would be hard to debug. I do not see why this would error out if there isnt something wrong. so I think we should throw the error here but we can revise this if we see genuine cases where ignoring is sensible.

Reply:
I really can see the benefits of both sides between failing fast vs. having to notice the warning post failure -- but I'll leave this up to you.